File: 110. Balanced Binary Tree/solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def isBalanced(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        balanced, height = self.myIsBalanced(root)
        return balanced
    
    def myIsBalanced(self, root):
        """
        :type root: TreeNode
        :rtype: bool, int
        """
        if not root:
            return True, 0
        else:
            lbal, lht = self.myIsBalanced(root.left)
            rbal, rht = self.myIsBalanced(root.right)
            mybal = (abs(lht - rht) <= 1) and lbal and rbal
            myht = 1 + max(lht, rht)
            return mybal, myht
        

File: 110. Balanced Binary Tree/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean isBalanced(TreeNode root) {
        Results result = checkBalanced(root);
        return result.getBalance();
    }
    
    public Results checkBalanced(TreeNode root) {
        Results results = new Results();
        
        if (root == null) {
            return results;
        }
        
        else if (root.left != null && root.right == null) {
            Results left = checkBalanced(root.left);
            results.setHeight(left.getHeight() + 1);
            results.setBalance(left.getBalance() && left.getHeight() <= 1);
            return results;
        } 
        
        else if (root.left == null && root.right != null) {
            Results right = checkBalanced(root.right);
            results.setHeight(right.getHeight() + 1);
            results.setBalance(right.getBalance() && right.getHeight() <= 1);
            return results;
        }
        
        else {
            Results left = checkBalanced(root.left);
            Results right = checkBalanced(root.right);
            
            results.setHeight(Math.max(right.getHeight(), left.getHeight()) + 1);
            results.setBalance(left.getBalance() &&
                               right.getBalance() &&
                               Math.abs(left.getHeight() - right.getHeight()) <= 1);
            return results;
        }
    }
    
    class Results {
        private boolean balance;
        private int height;
        
        public Results() {
            height = 0;
            balance = true;
        }
        
        public void setBalance(boolean balance) {
            this.balance = balance;
        }
        
        public boolean getBalance() {
            return balance;
        }
        
        public void setHeight(int height) {
            this.height = height;
        }
        
        public int getHeight() {
            return height;
        }
        
    }
}
